Java/core/src/main/java/com/amazon/randomcutforest/RandomCutForest.java

```diff
@@ -255,13 +246,75 @@ protected RandomCutForest(Builder<?> builder) {
         if (parallelExecutionEnabled) {
             // If the user specified a thread pool size, use it. Otherwise, use available
             // processors - 1.
-            threadPoolSize = builder.threadPoolSize.orElse(Runtime.getRuntime().availableProcessors() - 1);
             traversalExecutor = new ParallelForestTraversalExecutor(components, threadPoolSize);
-            updateExecutor = new ParallelForestUpdateExecutor(updateCoordinator, components, threadPoolSize);
+            updateExecutor = new ParallelForestUpdateExecutor(this.updateCoordinator, components, threadPoolSize);
         } else {
-            threadPoolSize = 0;
             traversalExecutor = new SequentialForestTraversalExecutor(components);
-            updateExecutor = new SequentialForestUpdateExecutor(updateCoordinator, components);
+            updateExecutor = new SequentialForestUpdateExecutor(this.updateCoordinator, components);
+        }
+    }
+
+    private void initCompact() {
+        PointStoreDouble pointStore = new PointStoreDouble(dimensions, sampleSize * numberOfTrees + 1);
+        IUpdateCoordinator<Integer> updateCoordinator = new PointStoreCoordinator(pointStore);
+        ComponentList<Integer> componentModels = new ComponentList<>(numberOfTrees);
+        for (int i = 0; i < numberOfTrees; i++) {
+            ITree<Integer> tree = new CompactRandomCutTreeDouble(sampleSize, rng.nextLong(), pointStore);
+            IStreamSampler<Integer> sampler = new CompactSampler(sampleSize, lambda, rng.nextLong(),
+                    storeSequenceIndexesEnabled);
+            componentModels.add(new SamplerPlusTree<>(sampler, tree));
+        }
+        this.updateCoordinator = updateCoordinator;
+        this.componentModels = componentModels;
+    }
+
+    private void initNoncompact() {
+        IUpdateCoordinator<double[]> updateCoordinator = new PassThroughCoordinator();
+        ComponentList<double[]> componentModels = new ComponentList<>(numberOfTrees);
+        for (int i = 0; i < numberOfTrees; i++) {
+            ITree<double[]> tree = RandomCutTree.builder().storeSequenceIndexesEnabled(storeSequenceIndexesEnabled)
+                    .centerOfMassEnabled(centerOfMassEnabled).randomSeed(rng.nextLong()).build();
+            IStreamSampler<double[]> sampler = new SimpleStreamSampler<>(sampleSize, lambda, rng.nextLong(),
+                    storeSequenceIndexesEnabled);
+            componentModels.add(new SamplerPlusTree<>(sampler, tree));
+        }
+        this.updateCoordinator = updateCoordinator;
+        this.componentModels = componentModels;
+    }
+
+    private RandomCutForest(Builder<?> builder) {
```

Comment:
question. why not keep this constructor as protected for the creation use case? then the other constructor will take three non-null parameters

Reply:
This constructor exists mostly for code organization, and all it does is set the RandomCutForest config variables from the builder. In particular, it doesn't set up models or executors. Since it doesn't create models, it doesn't really fully initialize the RCF state. Originally I wanted this to be a private method named `initConfig`, but it has to be a constructor because some of the forest instance variables are final.

Taking a step back, we want to be able to construct a forest in 2 ways:
* Create a forest from a config, where the models are initialized in the construcrtor
* Create a forest from a config and models that were initialized outside of the constructor. This is what we'll use during deserialization.

The current approach in this PR is to expose one constructor that takes a builder and model arguments. Then the decision to initialize models or not is based on whether the model arguments are null. I agree that it would be better two expose two constructors, one that takes model arguments and one that doesn't, but then we either have to duplicate the configuration initialization code, or else we have to make all the instance variables non-final so that they can be set in a method. Which solution do you like? Do you have different suggestion?

Reply:
To work around that a private method cannot set final members, how about a private constructor C which takes a config and an extra parameter (type and value doesn't matter) to differentiate from the config constructor A? C does the shared initialization logic. Constructors A and B call C and handle the rest.

Reply:
I like it, I'll include this change in the next revision.